pkg/cloud/services/eks/oidc.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package eks
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/service/eks"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/cluster-api/controllers/remote"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	trustPolicyConfigMapName      = "boilerplate-oidc-trust-policy"
+	trustPolicyConfigMapNamespace = metav1.NamespaceDefault
+
+	whitespaceRe = regexp.MustCompile(`(?m)[\t\n]`)
+)
+
+func (s *Service) reconcileOIDCProvider(cluster *eks.Cluster) error {
+	if !s.scope.ControlPlane.Spec.AssociateOIDCProvider || s.scope.ControlPlane.Status.OIDCProvider.ARN != "" {
+		return nil
+	}
+	s.scope.Info("Reconciling EKS OIDC Provider", "cluster-name", cluster.Name)
+	oidcProvider, err := s.CreateOIDCProvider(cluster)
+	if err != nil {
+		return errors.Wrap(err, "failed to create OIDC provider")
+	}
+	s.scope.ControlPlane.Status.OIDCProvider.ARN = oidcProvider
+	s.scope.ControlPlane.Status.OIDCProvider.TrustPolicy = whitespaceRe.ReplaceAllString(s.buildOIDCTrustPolicy(), "")
+	if err := s.scope.PatchObject(); err != nil {
+		return errors.Wrap(err, "failed to update control plane with OIDC provider ARN")
+	}
+
+	if err := s.reconcileTrustPolicy(); err != nil {
+		return errors.Wrap(err, "failed to reconcile trust policy in workload cluster")
+	}
+
+	return nil
+}
+
+func (s *Service) reconcileTrustPolicy() error {
+	ctx := context.Background()
+
+	clusterKey := client.ObjectKey{
+		Name:      s.scope.Name(),
+		Namespace: s.scope.Namespace(),
+	}
+
+	restConfig, err := remote.RESTConfig(ctx, s.scope.Client, clusterKey)
```

Comment:
We should create an issue to support having the template policy as a configmap on the management cluster along side the kubeconfig/cluster definition

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/issues/2089